### ISSUE TITLE
support new numpy complex types

### ISSUE THREAD

**Issue by tacaswell** [pydata/numexpr#458]
With numpy main compliing numexpr fails with (a lot) of errors that look like:

```
      numexpr/complex_functions.hpp: In function ‘void nc_assign(npy_cdouble*, npy_cdouble*)’:
      numexpr/complex_functions.hpp:36:6: error: request for member ‘real’ in ‘* r’, which is of non-class type ‘npy_cdouble’ {aka ‘__complex__ double’}
         36 |   r->real = x->real;
            |      ^~~~
      numexpr/complex_functions.hpp:36:16: error: request for member ‘real’ in ‘* x’, which is of non-class type ‘npy_cdouble’ {aka ‘__complex__ double’}
         36 |   r->real = x->real;
            |                ^~~~
```

I suspect the fix looks like https://github.com/scipy/scipy/pull/18975/ and is also a result of https://github.com/numpy/numpy/pull/24085 / https://github.com/numpy/numpy/issues/20880

**Comment by 27rabbitlt** [comment]
Can you please provide a snippet of code? I'm unfamiliar with this numpy's new feature T_T.

**Comment by tacaswell** [comment]
numpexpr fails to compile with the numpy default branch.

I think this change won't go out in stable release until numpy 2

**Comment by seberg** [comment]
It's not a new feature, the NumPy main branch just tries to use the C99/C++11 definitions for complex.  And those don't define `->real` and `->imag` so you need to replace it.

Since I bet you know that you have complex doubles, could probably just not use the NumPy types at all (so that you can use C++ `.real()` and `.imag()` no matter which NumPy it is) otherwise may need to cast or use some macro magic.

**Comment by 27rabbitlt** [comment]
thanks I’ll try to fix this

**Comment by tacaswell** [comment]
Closed by #461
